### DISCUSSION TITLE
Closing button

### DISCUSSION THREAD

**Discussion by charleselie** [slint-ui/slint#3562] [Q&A] [answered]
Hello,

I'am doing a graphic interface using slint with rust.
I want to do a frameless window, but now my problem is that I don't know how to close the window properly ...
is there any method, to simply close the window after a cliked event on a button ? 

thanks

**Comment by charleselie** [comment]
The way i'm currently doing it, is by getting an event from the ui and then using process::exit() in my rust code.
Is there a better way ?

**Reply by ogoffart** [reply]
I think process::exit is a perfectly valid way to do it.

I suggest to use Window::hide in the other comment I sent at the same time as your comment. But it's just a matter of taste or if you want to use multiple windows.

**Comment by ogoffart** [comment, marked as answer]
(Note: I've moved this issue to Discussions as it's better suited for general questions and community interactions. This helps keep our issue tracker focused on bug reports and feature requests)

The way to do what you want is to have a callback that closes the window from native code.  


```rust
fn main() {
    slint::slint! {
        import { Button } from "std-widgets.slint";
        export component Win inherits Window {
            callback close();
            Button { text: "Close"; clicked => { close(); }  }
        }
    }

    let win = Win::new().unwrap();
    let weak = win.as_weak();
    win.on_close(move || weak.unwrap().window().hide().unwrap());
    win.run().unwrap();
}

```

**Comment by charleselie** [comment]
Yes thanks for the feedback.

But I think that window::hide, doesn’t kill the process.

**Reply by ogoffart** [reply]
Window::hide will terminate the event loop if that's the last window

**Reply by Horbin-Magician** [reply]
If my program needs to constantly create new Windows, old Windows may be closed. If I use `hide` to close the windows, will it cause a problem of memory consumption? @ogoffart

**Reply by ogoffart** [reply]
> If I use hide to close the windows, will it cause a problem of memory consumption?

It shouldn't. The memory should be released if there are no handle to it.
One should be careful about Rc cycle though

**Reply by Horbin-Magician** [reply]
> It shouldn't. The memory should be released if there are no handle to it.

This means I need to release the handle to release the memory.
But handles need to be created in the UI thread, and at the same time can't be transferred between threads, so how do I release it without terminating the UI thread (since there may be other windows present)?

**Comment by StrongTheDev** [comment]
also this could work:

```rust
app.on_cls_win(|| {
    // std::process::exit(0);
    slint::quit_event_loop().unwrap();
});
```
